programs/lib/include/flexpret_noc.h

```diff
@@ -13,26 +14,83 @@
 #define NOC_TX_READY(val) (val & 0x01)
 #define NOC_DATA_AVAILABLE(val) (val & 0x02)
 
-// Blocking send word
-static void noc_send(uint32_t addr, uint32_t data) {
-    while (!NOC_TX_READY(NOC_CSR)) {}
-    NOC_DEST = addr;
-    NOC_DATA = data;
+/** Send a word over the NoC.
+ * Depending on the value of the timeout, the function will exhibit a different behavior:
+ *  * Blocking send is performed if timeout value is UNIT32_MAX,
+ *  * Non blocking send is performed if the timeout value is 0,
+ *  * And send within the given timeout, otherwise.
+ *
+ * Args:
+ *  * addr: id of the core to send to
+ *  * data: The data value to send
+ *  * timeout: the timeout in ns
+ * Returns:
+ *  * int: 1, if sending is successful, 0 otherwise
+ **/
+static int noc_send(uint32_t addr, uint32_t data, uint32_t timeout) {
+    if (timeout == UINT32_MAX) {
+        while (!NOC_TX_READY(NOC_CSR));
+        NOC_DEST = addr;
+        NOC_DATA = data;
+        return 1;
+    }
+    if (timeout == 0) {
+        if (NOC_TX_READY(NOC_CSR)) {
+            NOC_DEST = addr;
+            NOC_DATA = data;
+            return 1;
+        } else {
+            return 0;
+        }
+    }
+    uint32_t time = rdtime() + timeout;
+    while (rdtime() < time) {
+        if (NOC_TX_READY(NOC_CSR)) {
+            NOC_DEST = addr;
+            NOC_DATA = data;
+            return 1;
+        }
+    }
+    return 0;
 }
 
+
 // Send array, blocking
-static void noc_send_arr(uint32_t addr, uint32_t * data, int length) {
-    for (int i=0; i<length; i++) {
-        while (!NOC_TX_READY(NOC_CSR)) {}
-        NOC_DEST = addr; // FIXME: Can we get rid of these?
-        NOC_DATA = data[i];
-    }
+static void noc_send_arr(uint32_t addr, uint32_t *data, int length) {
+    for (int i = 0; i < length; i++)
+        noc_send(addr, data[i], UINT32_MAX);
+    
 }
 
-// Blocking read word
-static uint32_t noc_receive() {
-    while(!NOC_DATA_AVAILABLE(NOC_CSR)) {}
-    return NOC_DATA;
+
+/** Receive a word over the NoC.
+ * Depending on the value of the timeout, the function will exhibit a different 
+ * behavior:
+ *  * Blocking receive is performed if timeout value is UNIT32_MAX,
+ *  * Non blocking receive is performed if the timeout value is 0,
+ *  * And receive within the given timeout, otherwise.
+ *
+ * Args:
+ *  * timeout: the timeout in ns
+ * Returns:
+ *  * uint32_t: the received data, 0 otherwise
+ **/
+static uint32_t noc_receive(uint32_t timeout) {
+    if (timeout == UINT32_MAX) {
+        while (!NOC_DATA_AVAILABLE(NOC_CSR));
+        return NOC_DATA;
+    }
+    if (timeout == 0) {
+        if (NOC_DATA_AVAILABLE(NOC_CSR))
+            return NOC_DATA;
+        return 0;
+    }
+    uint32_t time = rdtime() + timeout;
+    while (rdtime() < time) {
+        if (NOC_DATA_AVAILABLE(NOC_CSR))
+            return NOC_DATA;
```

Comment:
This is open for discussion. But to improve predictability, it makes sense to me to make the `noc_receive()` constant-time by adding a `delay_until(time)` before returning. The current implementation enforces an upper bound, which could have timing variance if the function returns early. A `delay_until(time)` enforces the lower bound of execution time.

Reply:
I see your point, but I don't agree.
If we use `delay_until(timeout)`, then we may lose data if two cores send to the same core close enough in time (that is: both send happen at less than the timeout interval).
Better to catch up on the total execution time using `delay_until()` at the end of the task.
In addition, if the timeout will be equal to the trip duration for a message to arrive, then probably it is not worth it to put the hardware thread to sleep and then wake up.
Nevertheless, it is open for discussion, indeed.